File: android/app/src/test/kotlin/com/hananideas/batteryalarm/AppSettingsTest.kt
package com.hananideas.batteryalarm

import com.hananideas.batteryalarm.data.AppSettings
import org.junit.Assert.assertEquals
import org.junit.Test
import java.util.Locale

/**
 * Settings that reach the service from the method channel are untrusted input:
 * [AppSettings.sanitized] is the guard that keeps a bad value from turning into a
 * silent alarm or a 0-second repeat loop.
 */
class AppSettingsTest {

    @Test
    fun `defaults match the documented values`() {
        val defaults = AppSettings()

        assertEquals(10, defaults.alarmIntervalSeconds)
        assertEquals(80, defaults.alarmVolumePercent)
        assertEquals("en", defaults.language)
        assertEquals(1.0f, defaults.speechRate, 0.001f)
        assertEquals(1.0f, defaults.speechPitch, 0.001f)
    }

    @Test
    fun `an unsupported interval snaps to the nearest allowed value`() {
        assertEquals(15, AppSettings(alarmIntervalSeconds = 17).sanitized().alarmIntervalSeconds)
        assertEquals(5, AppSettings(alarmIntervalSeconds = 1).sanitized().alarmIntervalSeconds)
        assertEquals(60, AppSettings(alarmIntervalSeconds = 900).sanitized().alarmIntervalSeconds)
    }

    @Test
    fun `volume is clamped into an audible range`() {
        assertEquals(10, AppSettings(alarmVolumePercent = -20).sanitized().alarmVolumePercent)
        assertEquals(100, AppSettings(alarmVolumePercent = 400).sanitized().alarmVolumePercent)
    }

    @Test
    fun `speech rate and pitch are clamped`() {
        val sanitized = AppSettings(speechRate = 9f, speechPitch = 0.01f).sanitized()

        assertEquals(2.0f, sanitized.speechRate, 0.001f)
        assertEquals(0.5f, sanitized.speechPitch, 0.001f)
    }

    @Test
    fun `an empty announcement falls back to the default message`() {
        assertEquals(
            AppSettings.DEFAULT_MESSAGE,
            AppSettings(voiceMessage = "   ").sanitized().voiceMessage,
        )
    }

    @Test
    fun `an unknown language falls back to English`() {
        assertEquals("en", AppSettings(language = "kl").sanitized().language)
    }

    @Test
    fun `each supported language maps to a text-to-speech locale`() {
        assertEquals(Locale.ENGLISH, AppSettings(language = "en").ttsLocale())
        assertEquals("ur", AppSettings(language = "ur").ttsLocale().language)
        assertEquals("ar", AppSettings(language = "ar").ttsLocale().language)
        assertEquals("hi", AppSettings(language = "hi").ttsLocale().language)
    }

    @Test
    fun `a map from the method channel round-trips through the model`() {
        val original = AppSettings(
            userName = "Muhammad",
            voiceMessage = "{name}, unplug it.",
            alarmIntervalSeconds = 30,
            alarmVolumePercent = 65,
            flashEnabled = true,
            language = "ur",
        )

        assertEquals(original, AppSettings.fromMap(original.toMap()))
    }

    @Test
    fun `missing keys in an incoming map fall back to defaults`() {
        val parsed = AppSettings.fromMap(mapOf("userName" to "Ali"))

        assertEquals("Ali", parsed.userName)
        assertEquals(AppSettings().alarmIntervalSeconds, parsed.alarmIntervalSeconds)
        assertEquals(AppSettings().voiceMessage, parsed.voiceMessage)
    }

    @Test
    fun `an overlong name is truncated rather than rejected`() {
        val parsed = AppSettings(userName = "x".repeat(200)).sanitized()

        assertEquals(40, parsed.userName.length)
    }
}

File: test/features/settings/app_settings_test.dart
import 'package:battery_full_alarm/core/constants/app_constants.dart';
import 'package:battery_full_alarm/features/settings/domain/app_settings.dart';
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  group('AppSettings.sanitized', () {
    test('snaps an unsupported interval to the nearest allowed value', () {
      expect(
        const AppSettings(alarmIntervalSeconds: 17).sanitized().alarmIntervalSeconds,
        15,
      );
      expect(
        const AppSettings(alarmIntervalSeconds: 1).sanitized().alarmIntervalSeconds,
        5,
      );
      expect(
        const AppSettings(alarmIntervalSeconds: 900).sanitized().alarmIntervalSeconds,
        60,
      );
    });

    test('clamps the volume into an audible range', () {
      expect(const AppSettings(alarmVolumePercent: -5).sanitized().alarmVolumePercent, 10);
      expect(const AppSettings(alarmVolumePercent: 250).sanitized().alarmVolumePercent, 100);
    });

    test('clamps speech rate and pitch', () {
      const extreme = AppSettings(speechRate: 9, speechPitch: 0.01);
      expect(extreme.sanitized().speechRate, AppConstants.maxSpeechRate);
      expect(extreme.sanitized().speechPitch, AppConstants.minSpeechPitch);
    });

    test('restores the default message when the announcement is blank', () {
      expect(
        const AppSettings(voiceMessage: '   ').sanitized().voiceMessage,
        AppConstants.defaultVoiceMessage,
      );
    });

    test('trims and truncates an overlong name', () {
      final settings = AppSettings(userName: '  ${'x' * 200}  ').sanitized();
      expect(settings.userName.length, AppConstants.maxUserNameLength);
    });
  });

  group('serialisation', () {
    test('round-trips through the platform map', () {
      const original = AppSettings(
        userName: 'Muhammad',
        voiceMessage: '{name}, unplug it.',
        alarmIntervalSeconds: 30,
        alarmVolumePercent: 65,
        flashEnabled: true,
        language: AppLanguage.urdu,
        themeChoice: AppThemeChoice.dark,
      );

      expect(AppSettings.fromMap(original.toMap()), original);
    });

    test('falls back to defaults for missing keys', () {
      final parsed = AppSettings.fromMap(const {'userName': 'Ali'});

      expect(parsed.userName, 'Ali');
      expect(parsed.alarmIntervalSeconds, const AppSettings().alarmIntervalSeconds);
      expect(parsed.language, AppLanguage.english);
    });

    test('ignores values of the wrong type instead of throwing', () {
      final parsed = AppSettings.fromMap(const {
        'userName': 42,
        'voiceEnabled': 'yes',
        'alarmVolumePercent': 'loud',
      });

      expect(parsed.userName, '');
      expect(parsed.voiceEnabled, true);
      expect(parsed.alarmVolumePercent, const AppSettings().alarmVolumePercent);
    });

    test('the map keys match the ones Kotlin reads', () {
      // Guards against a rename on one side silently dropping a setting.
      expect(
        const AppSettings().toMap().keys.toSet(),
        {
          'userName',
          'voiceMessage',
          'voiceEnabled',
          'speechRate',
          'speechPitch',
          'language',
          'alarmIntervalSeconds',
          'alarmVolumePercent',
          'soundEnabled',
          'soundUri',
          'soundLabel',
          'vibrationEnabled',
          'flashEnabled',
          'notificationsEnabled',
          'autoStartMonitoring',
          'autoStartAfterBoot',
          'themeMode',
        },
      );
    });
  });

  group('enums', () {
    test('language codes map to locales', () {
      expect(AppLanguage.fromCode('ur'), AppLanguage.urdu);
      expect(AppLanguage.fromCode('ar').locale, const Locale('ar'));
      expect(AppLanguage.fromCode('nonsense'), AppLanguage.english);
    });

    test('theme codes map to Flutter theme modes', () {
      expect(AppThemeChoice.fromCode('dark').themeMode, ThemeMode.dark);
      expect(AppThemeChoice.fromCode(null).themeMode, ThemeMode.system);
    });
  });
}
